rfc/accepted/0021-command-driven-connectors.md

```diff
@@ -0,0 +1,119 @@
+# Command driven connectors
+
+- Feature Name: `command_driven_connectors`
+- Start Date: 2021-10-07
+- Tremor Issue: [tremor-rs/tremor-runtime#0000](https://github.com/tremor-rs/tremor-runtime/issues/0000)
+- RFC PR: [tremor-rs/tremor-rfcs#0000](https://github.com/tremor-rs/tremor-rfcs/pull/0000)
+
+## Summary
+[summary]: #summary
+
+This RFC introduces a control plane for connectors to allow controlling their behavior and actions via events.
+
+## Motivation
+[motivation]: #motivation
+
+There are many connectors, where, especially on the read side, there's no obvious "default" way to read data. For example object stores usually do not provide an API to stream changes (and if they did, the users might still want to read files in a different fashion). With command driven connectors, users will be able to read data that they need based on the commands they send to the connector.
+
+## Guide-level Explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Command-driven connectors define at least two ports - one for data and one for commands.
+As an example, let's look at a connector that reads files:
+
+```tremor
+define flow main
+flow
+  define connector file_connector from file
+    with 
+      codec="string",
+      config = {"command_driven": true}
+  end;
+  
+  define connector file_list from file
+    with 
+    codec = "json",
+    config = {
+      "path": "in.json",
+      "mode": "read", 
+    },
+  end;
+  
+  create pipeline main
+  pipeline
+    select { "command": "read", "path": event.path } from in into out;
+  end;
+  
+  create connector file_connector from file_connector;
+  create connector file_list from file_list;
+  
+  connect /connector/file_list/out to /pipeline/main/in;
+  
+  # This is the magic - we send the commands here, note the "control" port
+  connect /pipeline/main/out to /connector/file_connector/control;
+  
+  connect /connector/file_connector/data to /pipeline/main/out;
+end;
+```
+
+## Reference-level Explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Each command driven connector implements at least two ports - `data` and `control`.
+`control` is an input port, through which the commands are sent.
+Currently only reads are supported, so `data` is an output port.
+Each message in the `data` port is a single event, with metadata containing the original command.

```

Comment:
I think we could skip `single` here, and for the original command, the idea is to make the correlation between commands and their results easy. We could explore some alternatives here (maybe a `correlation` field which is an arbitrary Value that will be passed in the response meta?)

Reply:
Ah I see ja, perhaps if we go the rout of convenience functions they could create a `.correlation` key to the struct below that users can store in `state.whatever` and enrich their data from? Also lets think if we really need that but I think you're right we do. that would allow us to handle the correlation without additional information and template it out.

instead of users needing to remember i.e. the filename they can just store the correlation and it becomes irrelevant what the filename is


```tremor
 #I'll call it control plane correlation id `cpcid` because I'm lazy
  let control_event = file::open("myfile.txt");
  let state.cpcid= control_event .cpcid;
  emit control_event => "controlplane"

#... another day another event

let $cpcid = state.cpcid;
emit
```

Something like that I suppose?

Reply:
I think it makes sense to generate the correlation ID automagically, but also I think we should allow specifying the correlation ID, as I can imagine in some cases e.g. the filename is all that's needed. so maybe a good option here would be to have an optional parameter on the convenience functions, if it's not passed we generate an id?

Reply:
can you give an example where a user-supplied ID would be an advantage? I can't think of any but can think of a number where it'd be a possible problem most importantly that we loose all control over its behavior, shape, type etc.

Reply:
For example transferring files from one storage to another - you don't necessarily need to care about identifying where the file came from (so the ID isn't really helpful), but the filename/path is very useful.

Reply:
the name / path is already in metadata however, and the accounting of "this file is already open for write" is still required so I don't see much of a difference here, also it's extremely specific to files the RFC should work for any type of connector

Reply:
Ja and since the command / metadata is going to be an object if someone really wants to they can overwrite it as well but then it's clear "there be dragons" territory so we don't need to make any guarantees if that works / how that works if it keeps working in future releases :D